File: sorting/quicksort/quicksort.rb
def quicksort(array)
  # an array of zero or one elements is already sorted
  if array.size > 1
    # temporary arrays which will hold greater and lesser values
    less, more = [], []
    # select and remove a pivot value pivot from array
    pivot = array.pop
    array.each do |element|
      # Push the element to the appropriate temporary array, depending on the value
      (element <= pivot ? less : more).push element
    end
    # Update the array with the new order
    array.replace quicksort(less) + [pivot] + quicksort(more)
  end
  array
end

arr = [5,2,1,3,4]

puts "Before sort:"
puts arr.inspect

quicksort(arr)

puts "After sort:"
puts arr.inspect

File: sorting/quicksort/quicksort.js
function quicksort(array) {
  // An array of zero or one elements is already sorted
  if (array.length < 2) return array
  // Temporary arrays which will hold greater and lesser values
  var less = [], more = []
  // Select and remove a pivot value pivot from array
  var pivot = array.pop()
  // Push the element to the appropriate temporary array, depending on the value
  for (var i = 0, len = array.length; i < len; i++) 
    (array[i] <= pivot ? less : more).push(array[i])
  // Update the array with the new order
  array = [].concat(quicksort(less), [pivot], quicksort(more))
  return array
}

var array = [4,3,1,5,2]

console.log("Before sort:")
console.log(array)

array = quicksort(array)

console.log("After sort:")
console.log(array)